### DISCUSSION TITLE
AttributeError: module 'yolort' has no attribute 'utils'

### DISCUSSION THREAD

**Discussion by Wangwang99999** [zhiqwang/yolov5-rt-stack#423] [Q&A] [answered]
When I have finished build and install, and run export.py, it returns:
    import yolort.utils.dependency as _dependency
AttributeError: module 'yolort' has no attribute 'utils'

**Comment by zhiqwang** [comment]
### Install PyTorch and torchvision

Above all, follow the [official instructions](https://pytorch.org/get-started/locally/) to install PyTorch 1.8.0+ and torchvision 0.9.0+.

### Install yolort

We recommend to install yolort via pip:

```sh
python3 -m pip install yolort==0.6.3
```

to get the [latest release](https://github.com/zhiqwang/yolov5-rt-stack/archive/refs/tags/v0.6.3.zip) or

```
python3 -m pip install 'git+https://github.com/zhiqwang/yolov5-rt-stack.git'
```

to install the development version.

### Notes:
📝 Please try to use conda to manage your python libraries if possible.

📝 You can uninstall and reinstall `yolort` if you encounter something unexpected.

📝 And the following command will output a version message, which means that `yolort` has been installed correctly.
  ```sh
  python3 -c "import yolort;print(yolort.__version__)"
  ```

**Reply by Wangwang99999** [reply]
Thank you! I have installed with 'pip install yolort==0.6.3', but the problem is still the same.

**Reply by zhiqwang** [reply]
Hi @Wangwang99999 , Could you execute the following command and post the results?

```sh
python3 -c "import yolort;print(yolort.__version__)"
```

**Comment by Wangwang99999** [comment]
I run python3 -c "import yolort;print(yolort.\_\_version\_\_)"
It also trigger the problem of:
    import yolort.utils.dependency as _dependency
AttributeError: module 'yolort' has no attribute 'utils'

**Reply by zhiqwang** [reply]
Yep, This means that yolort is still not installed properly, can you try the following command instead:

```sh
python3 -m pip install yolort==0.6.3
```

**Reply by Wangwang99999** [reply]
More detailed information may help find problems:
```sh
D:\Python\ObjectDetection\yolov5-rt-stack>python3 -c "import yolort;print(yolort.__version__)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\__init__.py", line 4, in <module>
    from yolort import models
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\models\__init__.py", line 7, in <module>
    from .yolo import YOLO
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\models\yolo.py", line 8, in <module>
    from yolort.utils import load_state_dict_from_url
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\utils\__init__.py", line 15, in <module>
    from .image_utils import cv2_imshow, get_image_from_url, read_image_to_tensor
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\utils\image_utils.py", line 12, in <module>
    import yolort.utils.dependency as _dependency
AttributeError: module 'yolort' has no attribute 'utils'
```

**Reply by Wangwang99999** [reply]
It looks like fall into a circular import: yolort->models->image_utils->yolort

**Reply by zhiqwang** [reply]
> It looks like fall into a circular import: yolort->models->image_utils->yolort

Since you have downloaded the source code of yolort, in this case it is more appropriate to choose the installation path from the source code. But from the error message, the root cause is that pip and python are not mapped correctly. Maybe you can try the following command:

```sh
python3 -m pip install -e .
```

in the root directory of yolov5-rt-stack.

**Reply by Wangwang99999** [reply]
I run python3 -m pip install -e .
The problem is still the same:
import yolort.utils.dependency as _dependency
AttributeError: module 'yolort' has no attribute 'utils'

**Reply by zhiqwang** [reply]
That's weird, could you try the following and post the results
```sh
pip --version
pip3 --version
python --version
python3 --version
```

**Reply by zhiqwang** [reply]
And don't forget to uninstall yolort first when you install 
```sh
pip uninstall yolort
python3 -m pip uninstall yolort
```

**Reply by Wangwang99999** [reply]
Thank you very much!
pip --version and pip3 --version give the same result:
pip 21.1.3 from C:\Users\kka\AppData\Local\Programs\Python\Python36\lib\site-packages\pip (python 3.6)

python --version and python3 --version give the same result:
Python 3.6.5

**Reply by zhiqwang** [reply]
Thanks, seems pip is okay, that's weird, but there's something wrong with the installation, could you try to uninstall yolort and install yolort again.

**Reply by zhiqwang** [reply]
@Wangwang99999 , If the installation is correct, the following command will output a version message
```sh
python3 -c "import yolort;print(yolort.__version__)"
```

**Reply by Wangwang99999** [reply]
I have tried  sequentially the following install and uninstall instructions:
python3 setup.py build
python3 setup.py install
python3 export.py .....

pip3 uninstall yolort
pip install yolort==0.6.3
python3 export.py .....

pip3 uninstall yolort
python3 -m pip install yolort==0.6.3
python3 export.py .....

pip3 uninstall yolort
python3 -m pip install -e .
python3 export.py .....

but the problem is allways the same:
import yolort.utils.dependency as _dependency
AttributeError: module 'yolort' has no attribute 'utils'

**Reply by zhiqwang** [reply]
Could you try the following and post its result?
```
python3 -c "import yolort;print(yolort.__file__)"
```

**Reply by zhiqwang** [reply]
In general we don't recommend using the development version, so you don't need to download the entire yolov5-rt-stack repository. So delete the whole directory, it can confuse python library versions management. You can just download the file "export_model.py" from <https://github.com/zhiqwang/yolov5-rt-stack/blob/release/v0.6/tools/export_model.py>.

And please try to use conda to manage your python libraries if possible.

**Reply by Wangwang99999** [reply]
```sh
D:\Python\ObjectDetection\yolov5-rt-stack>python3 -c "import yolort;print(yolort.__file__)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\__init__.py", line 4, in <module>
    from yolort import models
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\models\__init__.py", line 7, in <module>
    from .yolo import YOLO
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\models\yolo.py", line 8, in <module>
    from yolort.utils import load_state_dict_from_url
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\utils\__init__.py", line 15, in <module>
    from .image_utils import cv2_imshow, get_image_from_url, read_image_to_tensor
  File "D:\Python\ObjectDetection\yolov5-rt-stack\yolort\utils\image_utils.py", line 12, in <module>
    import yolort.utils.dependency as _dependency
AttributeError: module 'yolort' has no attribute 'utils'
```

**Reply by Wangwang99999** [reply]
If a just download export_model.py, how to import other modules it depends?

**Reply by zhiqwang** [reply]
> If a just download export_model.py, how to import other modules it depends?

We publish yolort on PyPI. And that's the secrets of belows
```sh
python3 -m pip install yolort==0.6.3
```

**Reply by zhiqwang** [reply]
Delete the whole yolov5-rt-stack directory, uninstall and reinstall yolort from PyPI. And please try to use conda to manage your python libraries if possible.

**Comment by zhiqwang** [comment, marked as answer]
Hi @Wangwang99999 , Sorry for pointing in the wrong direction at first, this problem should be caused by the incompatibility of following writeup on some low versions of Windows system and Python.

https://github.com/zhiqwang/yolov5-rt-stack/blob/afe8d1009d17356723e45d26af94a38485e20d44/yolort/utils/image_utils.py#L12

We've fixed this problem in #427 , please update with the main branch and see if the latest update can help you. Feel free to open a new ticket if you have more questions.
